File: examples/go/type_assertions.go
//go:build ignore

package main

import "fmt"

// Go does not allow type assertions or type switches directly on type
// parameter values.  You must first convert to `any`, adding boilerplate

// Optional interfaces that items in a pipeline might implement.
type Validatable interface {
	Validate() error
}

type Loggable interface {
	LogEntry() string
}

// Process tries to validate and log a generic item.
// A type switch on the type-parameter value is illegal:
//
//	switch v := item.(type) {   // ERROR: cannot use type assertion on type parameter value
//	case Validatable: ...
//	case Loggable:    ...
//	}
//
// WORKAROUND: cast to `any` first, then use a type switch.
func Process[T any](item T) {
	// Type switch must go through `any`
	switch v := any(item).(type) {
	case Validatable:
		if err := v.Validate(); err != nil {
			fmt.Println("Validation failed:", err)
			return
		}
		fmt.Println("Validated OK")
	default:
		fmt.Println("No validation available")
	}

	// Same limitation for plain type assertions
	// entry, ok := item.(Loggable)       // ERROR
	entry, ok := any(item).(Loggable) // WORKAROUND
	if ok {
		fmt.Println("Log:", entry.LogEntry())
	} else {
		fmt.Println("Not loggable")
	}
}

// --- Test Implementations ---

// Order implements both Validatable and Loggable.
type Order struct {
	ID    int
	Total float64
}

func (o Order) Validate() error {
	if o.Total <= 0 {
		return fmt.Errorf("order %d has non-positive total", o.ID)
	}
	return nil
}

func (o Order) LogEntry() string {
	return fmt.Sprintf("Order#%d total=%.2f", o.ID, o.Total)
}

// RawEvent implements neither interface.
type RawEvent struct {
	Name string
}

func main() {
	fmt.Println("--- Order (implements both) ---")
	Process(Order{ID: 1, Total: 49.99})

	fmt.Println()
	fmt.Println("--- RawEvent (implements neither) ---")
	Process(RawEvent{Name: "click"})
}

File: examples/typescript/type_assertions.ts
export {};

// Go limitation: Go does not allow type assertions or type switches directly on type
// parameter values. You must first convert to `any`, adding boilerplate.

// TS: TS does not allow runtime type checks against interfaces, because
// interfaces are erased at compile time.

// Optional interfaces that items in a pipeline might implement.
interface Validatable {
  validate(): Error | null;
}

interface Loggable {
  logEntry(): string;
}

// A runtime instanceof check on the interface is illegal:
//
//   if (item instanceof Validatable) {   // ERROR: 'Validatable' only refers
//     ...                                // to a type, but is being used as
//   }                                    // a value here.
//
// WORKAROUND: use type-guard functions that perform structural runtime checks.
function isValidatable(v: unknown): v is Validatable {
  return typeof v === "object" && v !== null && "validate" in v;
}

function isLoggable(v: unknown): v is Loggable {
  return typeof v === "object" && v !== null && "logEntry" in v;
}

function process<T>(item: T): void {
  if (isValidatable(item)) {
    const err = item.validate();
    if (err !== null) {
      console.log("Validation failed:", err.message);
      return;
    }
    console.log("Validated OK");
  } else {
    console.log("No validation available");
  }

  // Same limitation for a plain type assertion
  // (item as Loggable).logEntry()        // compiles, but crashes at runtime if item is not Loggable
  //
  // WORKAROUND
  if (isLoggable(item)) {
    console.log("Log:", item.logEntry());
  } else {
    console.log("Not loggable");
  }
}

// --- Test Implementations ---

// Order implements both Validatable and Loggable.
class Order {
  constructor(
    public id: number,
    public total: number,
  ) {}

  validate(): Error | null {
    if (this.total <= 0) {
      return new Error(`order ${this.id} has non-positive total`);
    }
    return null;
  }

  logEntry(): string {
    return `Order#${this.id} total=${this.total.toFixed(2)}`;
  }
}

// RawEvent implements neither interface.
class RawEvent {
  constructor(public name: string) {}
}

function main() {
  console.log("--- Order (implements both) ---");
  process(new Order(1, 49.99));

  console.log();
  console.log("--- RawEvent (implements neither) ---");
  process(new RawEvent("click"));
}

main();
